Costfuzz tests using workloadreplay

rachitgsrivastava opened the pull request:
Creates a new costfuzz test that users workloadreplay to simulate customer queries/data for testing. Introduces a new interface called queryGenerator in query_comparison_util.go with a generate function that generalizes query generation across different tests(currently costfuzz.go and unoptimized_query_oracle.go) that utilizes query_comparison_util.go.

**Please note:**
We want to wait until process of passing credentials to tests is figured out - these include snowflake and gcs credentials. Without these credentials the process will fail.

We have tested workloadreplay in team city and it is ready to be looked at. @rytaft

rachitgsrivastava committed:
roachtest: workload replay test generation in costfuzz

Creates a new costfuzz test that users workloadreplay to simulate
customer queries/data for testing. Introduces a new interface
called queryGenerator in query_comparison_util.go with a generate
function that generalizes query generation across different
tests(currently costfuzz.go and unoptimized_query_oracle.go)
and utilizes query_comparison_util.go.

Epic CRDB-19060

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95390)
<!-- Reviewable:end -->

rytaft reviewed (commented):
Awesome! I know you're still figuring out some things with the credentials, but I figured I would leave some comments that you can address in parallel.

- It looks like you reverted some recent changes to these files, so you'll need to undo those changes (I think I identified most of them)
- I added a note to make sure you check that messages are getting sent to the correct logs
- Looks like you have some lint failures to address from the CI failures
- Otherwise, most of my comments are just nits

Great work!

> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-:-NMB-yZV1K-PWi27eWMc:b-ykyqh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rachitgsrivastava)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAE2jP6-CnrMw4eSxv):*
please rewrite and format commit message according to https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages
___
*[`pkg/cmd/roachtest/tests/costfuzz.go` line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAEKEYBc2fVvsmHfqK:-NMAEKEYBc2fVvsmHfqL:b-n3mwfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/cmd/roachtest/tests/costfuzz.go#L29)):*
> ```Go
> 
> func registerCostFuzz(r registry.Registry) {
> 	for _, setupName := range []string{"workloadReplay", sqlsmith.RandTableSetupName, sqlsmith.SeedMultiRegionSetupName} {
> ```

nit: let's follow the existing convention and call this "workload-replay"
___
*[`pkg/cmd/roachtest/tests/costfuzz.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAERPd8YpctDt1Izg0:-NMAERPd8YpctDt1Izg1:b-39soph) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/cmd/roachtest/tests/costfuzz.go#L34)):*
> ```Go
> 		switch setupName {
> 		case sqlsmith.SeedMultiRegionSetupName:
> 			clusterSpec = r.MakeClusterSpec(4, spec.Geo())
> ```

why did this change?
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAEa-AAOwyJigLBClc:-NMAEa-AAOwyJigLBCld:b-ti0323) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/cmd/roachtest/tests/query_comparison_util.go#L41)):*
> ```Go
> }
> 
> type queryGenerator interface {
> ```

nit: add a comment to explain what this is
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAEkeaBHwokgOftZto:-NMAEkeaBHwokgOftZtp:b-gwl4x2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/cmd/roachtest/tests/query_comparison_util.go#L44)):*
> ```Go
> 	Generate() string
> }
> ```

We typically add lines to assert that a struct implements a certain interface. I suggest adding the following:

```
var _ queryGenerator = (*workloadreplayGenerator)(nil)
```
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAF6nKEM7nDJBEj7Uu:-NMAF6nKEM7nDJBEj7Uv:b-2puj3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/cmd/roachtest/tests/query_comparison_util.go#L45)):*
> ```Go
> }
> 
> type workloadreplayGenerator struct {
> ```

nit: let's make the whole thing camel case (i.e., `workloadReplayGenerator`)
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAFEu6ENIIYEuUQ27M:-NMAFEu76Ei6dfg3Jgnu:b-y7ulju) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/cmd/roachtest/tests/query_comparison_util.go#L49)):*
> ```Go
> }
> 
> func (wlrg *workloadreplayGenerator) Generate() string {
> ```

nit: add comment `// Generate is part of the queryGenerator interface.`
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAFT-467W-QHLQJKCC:-NMAFT-467W-QHLQJKCD:b-wb3zgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/20b2e8167477b37c33042e72eb019350d8c26e78/pkg/cmd/roachtest/tests/query_comparison_util.go#L52)):*
> ```Go
> 	clusterCtx, clusterCancel = context.WithTimeout(ctx, t.Spec().(*registry.TestSpec).Timeout-2*time.Minute)
> 	defer clusterCancel()
> ```

I don't think you should remove this, and you probably should revert all rest of the changes to this function
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAFwiG5t-kvMhaV0q6:-NMAFwiG5t-kvMhaV0q7:b-vttp7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/cmd/roachtest/tests/query_comparison_util.go#L114)):*
> ```Go
> 	if qct.setupName == "workloadReplay" {
> 		logsuffix = "_workloadreplay"
> 	}
> ```

nit: I don't think this should be necessary. I think the logs will already exist in a separate directory labeled with workloadReplay
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAGuER7qWnFkjCSViL:-NMAGuER7qWnFkjCSViM:b9gwsg5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/cmd/roachtest/tests/query_comparison_util.go#L174)):*
> ```Go
> 			}
> 			fmt.Fprint(log, logType+":"+stmt)
> 			fmt.Fprint(log, "\n")
> ```

you should be careful about which file you're logging to. I think `log` is supposed to contain SQL statements that can be run directly in the SQL shell without modification, while `t.L()` and `failureLog` seem to be different (I'm not sure, though -- please confirm)

It's important for debugging purposes that we can completely recreate the sequence of statements that was run without doing any copy-pasting / modifying of the file.
___
*[`pkg/cmd/roachtest/tests/unoptimized_query_oracle.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMANXYtCRoMaalm676C:-NMANXYtCRoMaalm676D:bi5g01d) ([raw file](https://github.com/cockroachdb/cockroach/blob/20b2e8167477b37c33042e72eb019350d8c26e78/pkg/cmd/roachtest/tests/unoptimized_query_oracle.go#L49)):*
> ```Go
> 			switch setupName {
> 			case sqlsmith.SeedMultiRegionSetupName:
> 				clusterSpec = r.MakeClusterSpec(9, spec.Geo(), spec.GatherCores())
> ```

this shouldn't have changed
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMB-kiI8ckgw_rb9pRt:-NMB-kiI8ckgw_rb9pRu:b9r27l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/internal/workloadreplay/workloadreplay.go#L54)):*
> ```Go
> 	ctx, cancel := context.WithCancel(context.Background())
> 	return db, ctx, cancel
> ```

nit: remove blank line (here and all functions below)
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMANl70CRyaCoGfIqAx:-NMANl70CRyaCoGfIqAy:b-5r817j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/internal/workloadreplay/workloadreplay.go#L57)):*
> ```Go
> }
> 
> // writeRowsGCS writes data in rows to google cloud storage file. We write to gcs so we can import data from
> ```

nit: comments should be <= 80 columns wide
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAO4vZEtV4oGrnRev7:-NMAO4vZEtV4oGrnRev8:bsrp4qm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/internal/workloadreplay/workloadreplay.go#L99)):*
> ```Go
> 	rndNumber := rand.Float64()
> 	var query string
> 	// Randomly select between a non join or join query. nonjoinratio specifies bias based on how representive join
> ```

nit: long line
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAODiK3OMjN0DrGlNl:-NMAODiK3OMjN0DrGlNm:b-ilwr4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/internal/workloadreplay/workloadreplay.go#L101)):*
> ```Go
> 	// Randomly select between a non join or join query. nonjoinratio specifies bias based on how representive join
> 	// query bank cardinality is of join workloads relative to non join representation.
> 	if rndNumber <= nonjoinratio {
> ```

nit: nonjoinratio -> nonJoinRatio
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAOTgC8Gvw1n7sH0jc:-NMAOTgC8Gvw1n7sH0jd:b36a03r) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/internal/workloadreplay/workloadreplay.go#L142)):*
> ```Go
> 		finalStmt = m1.ReplaceAllString(finalStmt, d) // Standardize statement so that table name does not include
> 		// database or schema wild card eg. x.customers or .x.customers
> 		//should be replaced with customers
> ```

nit: fix comment formatting, and ideally put it above the relevant line
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 152 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMB-HL-8gzM2QrSoStl:-NMB-HL-8gzM2QrSoStm:bu7h2fr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/internal/workloadreplay/workloadreplay.go#L152)):*
> ```Go
> 
> // CreateRandomDataSnowflake create random data in snowflake using signatures that reflect customer workloads.
> // Write snowflake tables to GCS and return a mapping of table name to gcs filepath and schema.
> ```

nit: wrap long lines


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Changes look good! (next time, just respond to the comments you addressed by writing "Done." in Reviewable, which will resolve them)

> Reviewed 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-:-NOakmTB0aBoQKkK1tsb:b-q11wle)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rachitgsrivastava)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NOakPvr8TfDDp1cNREs):*
nit: commit messages should be prefixed with the main package that was changed. I'd change this to "roachtest: workload replay test generation in costfuzz"
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 46 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NOal4Hz1W3amCJvKr3k:-NOal4Hz1W3amCJvKr3l:bynf4h9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6df794a8d48c4480932c53531ae17e6aad88104/pkg/cmd/roachtest/tests/query_comparison_util.go#L46)):*
> ```Go
> }
> 
> // queryGenerator provides interface for tests to generate queries via method Generate which will vary per test.
> ```

nit: comments should be <= 80 columns wide.
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 97 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NOalPpK40CSpSNSeWHB:-NOalPpK40CSpSNSeWHC:b2ohhz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6df794a8d48c4480932c53531ae17e6aad88104/pkg/cmd/roachtest/tests/query_comparison_util.go#L97)):*
> ```Go
> 
> 	for i := 0; ; i++ {
> 		c.Start(clusterCtx, t.L(), option.DefaultStartOpts(), install.MakeClusterSettings())
> ```

Any reason this moved above the `if shouldExit()` block?


<!-- Sent from Reviewable.io -->

renatolabs commented on pkg/cmd/roachtest/tests/costfuzz.go:
```diff
@@ -26,7 +26,7 @@ import (
 
 func registerCostFuzz(r registry.Registry) {
-	for _, setupName := range []string{sqlsmith.RandTableSetupName, sqlsmith.SeedMultiRegionSetupName} {
+	for _, setupName := range []string{"workload-replay", sqlsmith.RandTableSetupName, sqlsmith.SeedMultiRegionSetupName} {
```
Reminder: set `RedactResults` here.

rickystewart commented:
FYI I have set up the build configuration. [Internal link](https://cockroachlabs.atlassian.net/browse/DEVINFHD-809?focusedCommentId=162525).

rachitgsrivastava committed:
roachtest: Adding environment variable GCE_PROJECT

Adding environment variable for GCE_PROJECT where nightlies artifacts will be written
to as well as the private gcs bucket path where nightlies artifacts/logs will be written
to.

Epic: none

Release note: None

jlinder commented on pkg/cmd/roachtest/tests/admission_control_elastic_backup.go:
```diff
@@ -92,9 +93,13 @@ func registerElasticControlForBackups(r registry.Registry) {
 					m.Go(func(ctx context.Context) error {
 						t.Status(fmt.Sprintf("during: creating full backup schedule to run every 20m (<%s)", time.Minute))
-						_, err := db.ExecContext(ctx,
+                                                gcsBackupTestingBucket := os.Getenv("BACKUP_TESTING_BUCKET")
+                                                if gcsBackupTestingBucket == "" {
+                                                        gcsBackupTestingBucket="cockroach-bucket-testing-private"
+                                                }
+                                                _, err := db.ExecContext(ctx,
```
The indentation here is too much. It should be the same indentation as line 95.

jlinder commented on pkg/cmd/roachtest/tests/admission_control_elastic_backup.go:
```diff
@@ -92,9 +93,13 @@ func registerElasticControlForBackups(r registry.Registry) {
-						_, err := db.ExecContext(ctx,
+                                                gcsBackupTestingBucket := os.Getenv("BACKUP_TESTING_BUCKET")
+                                                if gcsBackupTestingBucket == "" {
+                                                        gcsBackupTestingBucket="cockroach-bucket-testing-private"
```
The defaults should continue to be the `cockroachdb-backup-testing` bucket, not the new one. The new one should be what is configured in the environment variable.

jlinder commented on pkg/cmd/roachtest/tests/backup.go:
```diff
@@ -82,6 +82,11 @@ const (
 	rows3GiB   = rows30GiB / 10
 )
 
+backupTestingBucket := os.Getenv("BACKUP_TESTING_BUCKET")
+if backupTestingBucket == "" {
+        backupTestingBucket="cockroach-bucket-testing-private"
+}
```
Is there one central place where this can be checked and set instead of repeating it three times? Then the same variable can be used everywhere.

jlinder reviewed (commented):
A few comments on the commit making the backups bucket configurable.

Also, generally, every commit's message should follow the [commit guidelines](https://wiki.crdb.io/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages).

rytaft reviewed (changes requested):
@rachitgsrivastava This is looking good, but there are still a few comments that need to be addressed before this can be merged.

> Reviewed 8 of 8 files at r7, 1 of 3 files at r8, 8 of 8 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-:-NXS7hPL24mi-PB6mjmJ:b-fnn3sd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rachitgsrivastava)

___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NXS7XP42Wwf7I4i7ViK:-NXS7XP42Wwf7I4i7ViL:b-9tdbbh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2113f663bb726845d5aa5b54720aef4f01db2594/pkg/cmd/roachtest/tests/query_comparison_util.go#L132)):*
> ```Go
> 	logPath := filepath.Join(t.ArtifactsDir(), fmt.Sprintf("%s%03d.log", qct.name, iter))
> 	log, err := os.Create(logPath)
> 	fmt.Println(logPath)
> ```

looks like leftover debugging line


<!-- Sent from Reviewable.io -->

rachitgsrivastava reviewed (commented):
I have made the adjustments 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-:-NYZPRbKCDj2Oig45dSH:bsbzhvc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder, @michae2, @renatolabs, and @rytaft)

___
*[`pkg/cmd/roachtest/tests/admission_control_elastic_backup.go` line 98 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95390#gh-1190371933:-NXqVxcHCF_PebMdg3W0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eae04a3deca721ad92073392ab9384f4c8f82b2d/pkg/cmd/roachtest/tests/admission_control_elastic_backup.go#L98)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

The defaults should continue to be the `cockroachdb-backup-testing` bucket, not the new one. The new one should be what is configured in the environment variable.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/admission_control_elastic_backup.go` line 100 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95390#gh-1190370582:-NXqVxGw-RXPQCeAYvtz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eae04a3deca721ad92073392ab9384f4c8f82b2d/pkg/cmd/roachtest/tests/admission_control_elastic_backup.go#L100)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

The indentation here is too much. It should be the same indentation as line 95.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/backup.go` line 88 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95390#gh-1190373065:-NXqWAi32m9bl3_G_7ON:bl2tjyn) ([raw file](https://github.com/cockroachdb/cockroach/blob/eae04a3deca721ad92073392ab9384f4c8f82b2d/pkg/cmd/roachtest/tests/backup.go#L88)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Is there one central place where this can be checked and set instead of repeating it three times? Then the same variable can be used everywhere.
</blockquote></details>

I have changed this so in each file it is only performance once. If we want to reduce it so that is is called only one used between different files/packages that introduces some complexity which I do not want to block us right now.
___
*[`pkg/cmd/roachtest/tests/costfuzz.go` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NM0P8vL6pn4z2HZiVQj-r2-29:-NYZP0F3DPm5MjvVUSRY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6df794a8d48c4480932c53531ae17e6aad88104/pkg/cmd/roachtest/tests/costfuzz.go#L29)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Reminder: set `RedactResults` here.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAGuER7qWnFkjCSViL:-NXqYW78C_eYotNa1-zT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/cmd/roachtest/tests/query_comparison_util.go#L174)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

you should be careful about which file you're logging to. I think `log` is supposed to contain SQL statements that can be run directly in the SQL shell without modification, while `t.L()` and `failureLog` seem to be different (I'm not sure, though -- please confirm)

It's important for debugging purposes that we can completely recreate the sequence of statements that was run without doing any copy-pasting / modifying of the file.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 97 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NOalPpK40CSpSNSeWHB:-NXqRISLE1Hj6CF2fBiQ:b-5yhn6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6df794a8d48c4480932c53531ae17e6aad88104/pkg/cmd/roachtest/tests/query_comparison_util.go#L97)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Any reason this moved above the `if shouldExit()` block?
</blockquote></details>

I fixed this - just needed to rebase with master.
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NXS7XP42Wwf7I4i7ViK:-NXqYOxI3FPcjfd5sAMg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2113f663bb726845d5aa5b54720aef4f01db2594/pkg/cmd/roachtest/tests/query_comparison_util.go#L132)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

looks like leftover debugging line
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 1 of 9 files at r11, 11 of 12 files at r12, 7 of 7 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder, @michae2, @rachitgsrivastava, and @renatolabs)

___
*[`pkg/cmd/roachtest/tests/admission_control_elastic_backup.go` line 17 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NYjNtfWBwpZnfBni_kr:-NYjNtfWBwpZnfBni_ks:bf00vk) ([raw file](https://github.com/cockroachdb/cockroach/blob/df049d1a99a15f8ae6c1c99de041591c39bd3a07/pkg/cmd/roachtest/tests/admission_control_elastic_backup.go#L17)):*
> ```Go
> 	"fmt"
> 	"time"
>         "os"
> ```

nit: alignment is off
___
*[`pkg/cmd/roachtest/tests/costfuzz.go` line 48 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NYeuDK-4fNn4hbyi-VR:-NYeuDK07p1kyhczgXG4:brshk8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e4952287c6d94f0aefc0d8e2dc216e7cf23cd21/pkg/cmd/roachtest/tests/costfuzz.go#L48)):*
> ```Go
> 			Owner:           registry.OwnerSQLQueries,
> 			Timeout:         timeOut,
> 			RedactResults: redactResults,
> ```

I think the linter is going to complain about this alignment
___
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 957 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NYjNOqlD5nU1Ig3TtmY:-NYjNOqlD5nU1Ig3TtmZ:b-fjtwog) ([raw file](https://github.com/cockroachdb/cockroach/blob/df049d1a99a15f8ae6c1c99de041591c39bd3a07/pkg/cmd/roachtest/tests/mixed_version_backup.go#L957)):*
> ```Go
> 		gcsBackupTestingBucket = "cockroachdb-backup-testing"
> 	}
> 	return fmt.Sprintf("gs://" + gcsBackupTestingBucket  + "/%s_%s?AUTH=implicit", bc.name, bc.nonce)
> ```

Looks like the `mixed-version` text got removed from the path
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAGuER7qWnFkjCSViL:-NYjN0CW6MHS_tr8DNpr:bc4fyf6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/cmd/roachtest/tests/query_comparison_util.go#L174)):*
<details><summary><i>Previously, rachitgsrivastava wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Doesn't look like anything changed here. Did you check that these are the right files?
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 97 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NOalPpK40CSpSNSeWHB:-NYjMjIZ1HSjhzRi_7jc:b-vmik6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6df794a8d48c4480932c53531ae17e6aad88104/pkg/cmd/roachtest/tests/query_comparison_util.go#L97)):*
<details><summary><i>Previously, rachitgsrivastava wrote…</i></summary><blockquote>

I fixed this - just needed to rebase with master.
</blockquote></details>

Doesn't look fixed to me.
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NXS7XP42Wwf7I4i7ViK:-NYjMnLI2aYYt3nQ0bwd:boof3vr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2113f663bb726845d5aa5b54720aef4f01db2594/pkg/cmd/roachtest/tests/query_comparison_util.go#L132)):*
<details><summary><i>Previously, rachitgsrivastava wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Still there.


<!-- Sent from Reviewable.io -->

rachitgsrivastava commented:
*[`pkg/cmd/roachtest/tests/costfuzz.go` line 48 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NYeuDK-4fNn4hbyi-VR:-NYyHAjm5xyiogzGeNGJ:b20i1q9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e4952287c6d94f0aefc0d8e2dc216e7cf23cd21/pkg/cmd/roachtest/tests/costfuzz.go#L48)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think the linter is going to complain about this alignment
</blockquote></details>

Fixed the indentation


<!-- Sent from Reviewable.io -->

rachitgsrivastava commented:
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAGuER7qWnFkjCSViL:-NYyMvb8De8HI2ag8uau:b5lnstt) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/cmd/roachtest/tests/query_comparison_util.go#L174)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Doesn't look like anything changed here. Did you check that these are the right files?
</blockquote></details>

t.L() is writing to the build logs and is more to track progress when running the test. All steps we want to reproduce eg. signature, path on gcs of data generated, query that we are running and if the query is valid all are logged in the artifacts directory. fmt.* in logTest writes to costfuzz*.log in artifacts so I am seeing the logging I expect in both sets of logs.


<!-- Sent from Reviewable.io -->

rachitgsrivastava commented:
*[`pkg/cmd/roachtest/tests/admission_control_elastic_backup.go` line 17 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NYjNtfWBwpZnfBni_kr:-NYyPflo9bRPduPjwoeX:b1b7bz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/df049d1a99a15f8ae6c1c99de041591c39bd3a07/pkg/cmd/roachtest/tests/admission_control_elastic_backup.go#L17)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: alignment is off
</blockquote></details>

Fixed this.


<!-- Sent from Reviewable.io -->

rachitgsrivastava commented:
*[`pkg/cmd/roachtest/tests/mixed_version_backup.go` line 957 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NYjNOqlD5nU1Ig3TtmY:-NYyR5jE0jIX71c-WO8H:be4zdwr) ([raw file](https://github.com/cockroachdb/cockroach/blob/df049d1a99a15f8ae6c1c99de041591c39bd3a07/pkg/cmd/roachtest/tests/mixed_version_backup.go#L957)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Looks like the `mixed-version` text got removed from the path
</blockquote></details>

I have fixed this path


<!-- Sent from Reviewable.io -->

rachitgsrivastava commented:
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 97 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NOalPpK40CSpSNSeWHB:-NYyR9Fx9Yu4ayl6YSfv:b-nrb2ua) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6df794a8d48c4480932c53531ae17e6aad88104/pkg/cmd/roachtest/tests/query_comparison_util.go#L97)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Doesn't look fixed to me.
</blockquote></details>

Sorry I do not see the disparity here between this and the original code. What part are you referring to?


<!-- Sent from Reviewable.io -->

rachitgsrivastava commented:
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NXS7XP42Wwf7I4i7ViK:-NYyRICHChYMyKkMLBk-:bdkbvg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2113f663bb726845d5aa5b54720aef4f01db2594/pkg/cmd/roachtest/tests/query_comparison_util.go#L132)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Still there.
</blockquote></details>

Sorry which part are you referring to here? 


<!-- Sent from Reviewable.io -->

rachitgsrivastava reviewed (commented):
I have reviewed changes and left some questions for two of the comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-:-NYyRQJZ9ZOdlwsqwNkA:b-9ozyey)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder, @michae2, @renatolabs, and @rytaft)



<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Thanks, responded below

> Reviewed 7 of 7 files at r14, 7 of 7 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-:-NYjLyKL3ZZrTDh-D8XN:b-q0oqql)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder, @michae2, @rachitgsrivastava, and @renatolabs)

___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NMAGuER7qWnFkjCSViL:-NYzOjzo9ovjmYqYgQko:b-gsh0ql) ([raw file](https://github.com/cockroachdb/cockroach/blob/8677076b194dedc759cd247f66fd6ce315fb1f98/pkg/cmd/roachtest/tests/query_comparison_util.go#L174)):*
<details><summary><i>Previously, rachitgsrivastava wrote…</i></summary><blockquote>

t.L() is writing to the build logs and is more to track progress when running the test. All steps we want to reproduce eg. signature, path on gcs of data generated, query that we are running and if the query is valid all are logged in the artifacts directory. fmt.* in logTest writes to costfuzz*.log in artifacts so I am seeing the logging I expect in both sets of logs.
</blockquote></details>

Thanks for checking
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 97 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NOalPpK40CSpSNSeWHB:-NYzOz-Q5a6bvtiCTub8:b-218ww6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6df794a8d48c4480932c53531ae17e6aad88104/pkg/cmd/roachtest/tests/query_comparison_util.go#L97)):*
<details><summary><i>Previously, rachitgsrivastava wrote…</i></summary><blockquote>

Sorry I do not see the disparity here between this and the original code. What part are you referring to?
</blockquote></details>

You moved `c.Start(clusterCtx, t.L(), option.DefaultStartOpts(), install.MakeClusterSettings())` above the `if shouldExit() {` block, and it's not clear why.
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NXS7XP42Wwf7I4i7ViK:-NYzOr5520NVEvWtbM0j:b-tj97ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/2113f663bb726845d5aa5b54720aef4f01db2594/pkg/cmd/roachtest/tests/query_comparison_util.go#L132)):*
<details><summary><i>Previously, rachitgsrivastava wrote…</i></summary><blockquote>

Sorry which part are you referring to here? 
</blockquote></details>

`fmt.Println(logPath)`


<!-- Sent from Reviewable.io -->

rytaft commented:
Also, your build is failing. Looks like you have some linter failures, etc.





<!-- Sent from Reviewable.io -->

rytaft commented:
Please also update the PR message on GitHub to follow guidelines as well (and remove parts that are no longer relevant)





<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder, @michae2, @rachitgsrivastava, and @renatolabs)

___
*[`-- commits` line 11 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NYzPpz1BbY7gTZfPMrd):*
nit: This commit doesn't use the correct style. As I mentioned previously, commit messages should be prefixed with the main package that was changed. Ideally this commit would also include a bit more information in the commit message about the purpose of this new env variable, and it needs a release note (probably `Release note: None`).


<!-- Sent from Reviewable.io -->

rachitgsrivastava commented:
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 97 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NOalPpK40CSpSNSeWHB:-NZ6av2yBTieKYH5mIk-:bx8eoka) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6df794a8d48c4480932c53531ae17e6aad88104/pkg/cmd/roachtest/tests/query_comparison_util.go#L97)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

You moved `c.Start(clusterCtx, t.L(), option.DefaultStartOpts(), install.MakeClusterSettings())` above the `if shouldExit() {` block, and it's not clear why.
</blockquote></details>

thanks for the catch that happened with some copy and pasting - i have fixed the order.


<!-- Sent from Reviewable.io -->

rachitgsrivastava commented:
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 132 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NXS7XP42Wwf7I4i7ViK:-NZ6bGhaAuiZaWYf0-dk:bvdltxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2113f663bb726845d5aa5b54720aef4f01db2594/pkg/cmd/roachtest/tests/query_comparison_util.go#L132)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

`fmt.Println(logPath)`
</blockquote></details>

got rid of it


<!-- Sent from Reviewable.io -->

rachitgsrivastava commented:
*[`-- commits` line 11 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NYzPpz1BbY7gTZfPMrd):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: This commit doesn't use the correct style. As I mentioned previously, commit messages should be prefixed with the main package that was changed. Ideally this commit would also include a bit more information in the commit message about the purpose of this new env variable, and it needs a release note (probably `Release note: None`).
</blockquote></details>

I have changed the commit message.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> once you fix the linter errors. Looks like you also need to run `./dev gen bazel`.

Might be good for @jlinder to sign off too

> Reviewed 8 of 8 files at r16, 7 of 7 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-:-NZ8id4y1ubZ3q-LDroO:b-tpsjca)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jlinder, @michae2, @rachitgsrivastava, and @renatolabs)

___
*[`pkg/cmd/roachtest/tests/costfuzz.go` line 40 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZ8iXX--lsiiRjt4UI1:-NZ8iXX--lsiiRjt4UI2:b-7u9t7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/cmd/roachtest/tests/costfuzz.go#L40)):*
> ```Go
> 			clusterSpec = r.MakeClusterSpec(1)
> 			timeOut = time.Hour * 2
>       redactResults = true
> ```

nit: fix indentation


<!-- Sent from Reviewable.io -->

renatolabs reviewed (changes requested):
Most important comment is the one about use of `log.Fatal`, we shouldn't use that in tests. I'm happy to work together on the other comments if you like, just ping me on Slack.

> Reviewed 4 of 12 files at r12, 8 of 8 files at r16.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-:-NZX1Vwr0HjQVsnrmksp:b59169o)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jlinder, @michae2, and @rachitgsrivastava)

___
*[`pkg/cmd/roachtest/tests/admission_control_elastic_backup.go` line 101 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWT41h30NmDrAv4uFo:-NZWT41i7qoNoBJRU2uG:b-ppfujr) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/cmd/roachtest/tests/admission_control_elastic_backup.go#L101)):*
> ```Go
> 						if gcsBackupTestingBucket == "" {
> 							gcsBackupTestingBucket="cockroachdb-backup-testing"
> 						}
> ```

Maybe I'm missing something, but why do we need to allow overwriting the bucket on non-private tests?
___
*[`pkg/cmd/roachtest/tests/admission_control_elastic_backup.go` line 101 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWVJkX3-7m_1fQjYnJ:-NZWVJkX3-7m_1fQjYnK:b26b3rp) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/cmd/roachtest/tests/admission_control_elastic_backup.go#L101)):*
> ```Go
> 						if gcsBackupTestingBucket == "" {
> 							gcsBackupTestingBucket="cockroachdb-backup-testing"
> 						}
> ```

If we do want to use the private bucket everywhere, we can reference the `backupTestingBucket` variable directly instead.
___
*[`pkg/cmd/roachtest/tests/backup.go` line 87 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWTCA47FKDWWJNEBLE:-NZWTCA47FKDWWJNEBLF:b-4icabm) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/cmd/roachtest/tests/backup.go#L87)):*
> ```Go
> var backupTestingBucket string
> 
> func init() {
> ```

It's preferable to not have an `init` function on this package: it's shared across hundreds of tests. In this case, it's sufficient to have something like:

```go
var backupTestingBucket = func() string {
    if env := os.Getenv("BACKUP_TESTING_BUCKET"); env != "" {
        return env
    }

    return "cockroachdb-backup-testing"
}()
```
___
*[`pkg/cmd/roachtest/tests/costfuzz.go` line 31 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWSPj9ArhhBlublfHw:-NZWSPj9ArhhBlublfHx:b-pacgj4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4950539260088f1370ecd7dc99b58e310eed298a/pkg/cmd/roachtest/tests/costfuzz.go#L31)):*
> ```Go
> 	for _, setupName := range []string{"workload-replay", sqlsmith.RandTableSetupName, sqlsmith.SeedMultiRegionSetupName} {
> 		setupName := setupName
> 		redactResults := false
> ```

Probably safer to adopt a "private by default" approach by setting this to `true` and setting it back to `false` on tests that can be public. This should cover us when we add a new test case here, in case it's private and we forget to set the appropriate flag.
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 198 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZX1fZL0j4_ZFWcSAtg:-NZX1fZL0j4_ZFWcSAth:bw18v7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/cmd/roachtest/tests/query_comparison_util.go#L198)):*
> ```Go
> 		for {
> 			t.L().Printf("Choosing Random Query")
> 			finalStmt, signatures = workloadreplay.ChooseRandomQuery(log)
> ```

`ChooseRandomQuery` and `CreateRandomDataSnowflake` are creating a new connections to Snowflake each time they are called, and these functions are being executed in this "infinite" loop. Probably worth keeping a handle to the connection and reusing it.

Also, I don't see where the connection is closed, maybe I missed something.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 21 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWWMIqD2mwM4t_JS_D:-NZWWMIqD2mwM4t_JS_E:b5larek) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L21)):*
> ```Go
> 
> const (
> 	projectID    = "e2e-infra-381422"
> ```

This constant doesn't seem to be used. Maybe the accountID used down below should be made into a constant instead?
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 52 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWWbXF87xD_ChQhZSm:-NZWWbXF87xD_ChQhZSn:b-eweye3) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L52)):*
> ```Go
> 	db, err := sql.Open("snowflake", dsn)
> 	if err != nil {
> 		log.Fatalf("Could not open sql connection %v", err)
> ```

`log.Fatal` will cause the entire process to exit abruptly (`os.Exit(1)`). No cleanup will take place (i.e., clusters running tests will continue to exist), and all other roachtests will not run in the nightly build.

Calls to `log.Fatalf` in this file should return an error instead and propagate that to the framework so that this is marked and reported as a regular test failure.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 54 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWYOOM-U70Gk0Uu5Vt:-NZWYOOM-U70Gk0Uu5Vu:b7l4vll) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L54)):*
> ```Go
> 		log.Fatalf("Could not open sql connection %v", err)
> 	}
> 	ctx, cancel := context.WithCancel(context.Background())
> ```

This function should not return a new context. The context passed to the functions below should be the same as the context passed to the test. This guarantees that, for example, if the test times out, the functions that take that context return appropriately.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 82 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWZo914vYtLPA_J21W:-NZWZo914vYtLPA_J21X:ba44rx9) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L82)):*
> ```Go
> 	wc.ContentType = "text/plain"
> 
> 	cols, _ := rows.Columns()
> ```

Missing error check.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 83 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWZpSB1ysRjYE0cOcO:-NZWZpSB1ysRjYE0cOcP:bosdpf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L83)):*
> ```Go
> 
> 	cols, _ := rows.Columns()
> 	cName := strings.Join(cols[:], ",")
> ```

Nit: unnecessary `[:]`.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 93 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWZzHe6wezfCUBzEJV:-NZWZzHe6wezfCUBzEJW:bosdpf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L93)):*
> ```Go
> 	for rows.Next() {
> 		_ = rows.Scan(colPointers...)
> 		rowData := strings.Join(colContainer[:], ",")
> ```

Nit: unnecessary `[:]`.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 109 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZW_UWPD8jluzXDRHj8:-NZW_UWPD8jluzXDRHj9:b-26pf6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L109)):*
> ```Go
> 
> func ChooseRandomQuery(log *os.File) (string, map[string][]string) {
> 	rndNumber := rand.Float64()
> ```

Food for thought: should we be using a `rand` from a source we control (instead of the global `rand`). In other words, using something like

```go
rng := rand.New(rand.NewSource(someSeed))

func ChooseRandomQuery(rng *rand.Rand) {
    // ... 
    rndNumber := rng.Float64()
}
```

This would allow us to reproduce failures by passing a specific seed and guarantee `ChooseRandomQuery` will generate the exact same queries as the ones picked in the failed test.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 120 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWa4flAc94nxpHKR4A:-NZWa4flAc94nxpHKR4B:bghlqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L120)):*
> ```Go
> 	} else {
> 		fmt.Fprint(log, "JOIN QUERY")
> 		query = "select a.statement,replace(regexp_replace(lower(replace(statement,'\"','')),concat('[\\\\\\S]*\\\\\\.[\\\\\\S]*\\\\\\.',lower(b.table_name)),lower(b.table_name)),'..','') statement_2,ifnull(a.database_name,'') database_name,a.CID,a.TID,a.schema_name,b.table_name,b.table_signature,b.cockroachdb_schema cockroachdb_schema from (select * from (select * from (select *,rank() over (partition by statement,schema_name,database_name,cid,tid order by run_date desc) as rnk from (select * from DATA_ENG.TEST_CONFIG.QUERYBANK_SELECT where disqualify=0)) where rnk=1)  sample (1 rows))  a inner join (select * from (select *,rank() over (partition by table_signature,schema_name,database_name,cid,tid  order by run_date desc) as rnk from DATA_ENG.TEST_CONFIG.TABLE_SIGNATURES) where rnk=1) b on a.SCHEMA_NAME=b.SCHEMA_NAME and contains(lower(statement),lower(table_name)) "
> ```

Might be worth formatting these queries to make them more palatable and easier to adapt when we have to.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 135 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWaDb3Af6Jljxr5PDR:-NZWaDb3Af6Jljxr5PDS:b3j6bl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L135)):*
> ```Go
> 	signatures := make(map[string][]string)
> 	var finalStmt string
> 	finalStmt = ""
> ```

Nit: this is not needed, `finalStmt` is already initialized with the empty string.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 149 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWagtN4CATvMuoR5gq:-NZWagtN4CATvMuoR5gr:b-evwhfh) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L149)):*
> ```Go
> 		}
> 		if finalStmt == "" {
> 			finalStmt = strings.ReplaceAll(strings.ToLower(stmt), string('"'), string(""))
> ```

Nit:

```go
strings.ReplaceAll(strings.ToLower(stmt), `"`, "")
```

reads a little better.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 179 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZX0frKC2-DodwAtZmb:-NZX0frKC2-DodwAtZmc:b-o858e) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L179)):*
> ```Go
> 		_, err := db.QueryContext(ctx, signatureSchema[0])
> 		if err != nil {
> 			fmt.Fprint(log, "COULD NOT GENERATE DATASET - signature execution failed:"+signatureSchema[0])
> ```

It would probably make our lives easier to use `log.Logger` instances here rather than operating on `*os.File`s directly.


<!-- Sent from Reviewable.io -->

rachitgsrivastava reviewed (commented):
Made some changes and left some comments. Lets sync Friday on this @renatolabs 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-:-N_GTCIk4VNSjPMj1Okp:bsa9bkp)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jlinder, @michae2, @renatolabs, and @rytaft)

___
*[`pkg/cmd/roachtest/tests/admission_control_elastic_backup.go` line 101 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWT41h30NmDrAv4uFo:-N_FqOa19uFsHFY4W7mU:bi11cqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/cmd/roachtest/tests/admission_control_elastic_backup.go#L101)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Maybe I'm missing something, but why do we need to allow overwriting the bucket on non-private tests?
</blockquote></details>

This was a design decision @jlinder suggested where if the env is not not set we continue to use the old bucket by default. I am assuming for other processes? 
___
*[`pkg/cmd/roachtest/tests/admission_control_elastic_backup.go` line 101 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWVJkX3-7m_1fQjYnJ:-N_FqnaHDW0pNF9AzNYZ:b4jtj9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/cmd/roachtest/tests/admission_control_elastic_backup.go#L101)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

If we do want to use the private bucket everywhere, we can reference the `backupTestingBucket` variable directly instead.
</blockquote></details>

sorry what variable are you talking about here?
___
*[`pkg/cmd/roachtest/tests/backup.go` line 87 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWTCA47FKDWWJNEBLE:-N_9IfFF519kp2C0jSSw:bujqf29) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/cmd/roachtest/tests/backup.go#L87)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

It's preferable to not have an `init` function on this package: it's shared across hundreds of tests. In this case, it's sufficient to have something like:

```go
var backupTestingBucket = func() string {
    if env := os.Getenv("BACKUP_TESTING_BUCKET"); env != "" {
        return env
    }

    return "cockroachdb-backup-testing"
}()
```
</blockquote></details>

acknowledged and change made
___
*[`pkg/cmd/roachtest/tests/costfuzz.go` line 31 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWSPj9ArhhBlublfHw:-N_AH8Y65kVBqcApWDZA:bp8n1dx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4950539260088f1370ecd7dc99b58e310eed298a/pkg/cmd/roachtest/tests/costfuzz.go#L31)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Probably safer to adopt a "private by default" approach by setting this to `true` and setting it back to `false` on tests that can be public. This should cover us when we add a new test case here, in case it's private and we forget to set the appropriate flag.
</blockquote></details>

It seems like most of the tests so far are 'public' and workloadreplay was the first "private" one....so not sure if we want that- also it seems like the default value is False - so if we want "private" to be default(I am not saying we should but assuming this is the case) should that not be changed wherever the default value is set?
___
*[`pkg/cmd/roachtest/tests/query_comparison_util.go` line 198 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZX1fZL0j4_ZFWcSAtg:-N_Ev4na-sva0XW-WvVV:b-cib1sx) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/cmd/roachtest/tests/query_comparison_util.go#L198)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

`ChooseRandomQuery` and `CreateRandomDataSnowflake` are creating a new connections to Snowflake each time they are called, and these functions are being executed in this "infinite" loop. Probably worth keeping a handle to the connection and reusing it.

Also, I don't see where the connection is closed, maybe I missed something.
</blockquote></details>

From https://pkg.go.dev/database/sql#Open 

"The returned DB is safe for concurrent use by multiple goroutines and maintains its own pool of idle connections. "

" It is rarely necessary to close a DB."
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 21 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWWMIqD2mwM4t_JS_D:-N_ANjKi-NLWSGGuZyGm:b8ad2mw) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L21)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

This constant doesn't seem to be used. Maybe the accountID used down below should be made into a constant instead?
</blockquote></details>

acknowledged and changed
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 52 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWWbXF87xD_ChQhZSm:-N_AcqNe-vPg8cabxc16:b-h5x8l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L52)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

`log.Fatal` will cause the entire process to exit abruptly (`os.Exit(1)`). No cleanup will take place (i.e., clusters running tests will continue to exist), and all other roachtests will not run in the nightly build.

Calls to `log.Fatalf` in this file should return an error instead and propagate that to the framework so that this is marked and reported as a regular test failure.
</blockquote></details>

ack that that log.Fatal should not be used here...I have fixed this. However, this should not be a test failure as there can be a temp. snowflake connection issue which may not occur next time a connection is made. Change has been made for logging and not abruptly exiting.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 54 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWYOOM-U70Gk0Uu5Vt:-N_FVrfCE4PxdTAMHCgK:b-fgoim3) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L54)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

This function should not return a new context. The context passed to the functions below should be the same as the context passed to the test. This guarantees that, for example, if the test times out, the functions that take that context return appropriately.
</blockquote></details>

I agree and have changed this.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 82 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWZo914vYtLPA_J21W:-N_AfBGK-Qh10x9XrQ6z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L82)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Missing error check.
</blockquote></details>

Done.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 83 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWZpSB1ysRjYE0cOcO:-N_AfH2O82L1dlKUZGba:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L83)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nit: unnecessary `[:]`.
</blockquote></details>

fixed
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 109 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZW_UWPD8jluzXDRHj8:-N_FVE1RAnckTM6gvhZT:b1tao88) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L109)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Food for thought: should we be using a `rand` from a source we control (instead of the global `rand`). In other words, using something like

```go
rng := rand.New(rand.NewSource(someSeed))

func ChooseRandomQuery(rng *rand.Rand) {
    // ... 
    rndNumber := rng.Float64()
}
```

This would allow us to reproduce failures by passing a specific seed and guarantee `ChooseRandomQuery` will generate the exact same queries as the ones picked in the failed test.
</blockquote></details>

Seeding from a source we control will not matter if the snowflake data changes over time. 
We store the data(tables we are querying from) and queries on gcs to reproduce results.
___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 179 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZX0frKC2-DodwAtZmb:-N_FpnNX3Xzph68ZGZeP:b20s1af) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L179)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

It would probably make our lives easier to use `log.Logger` instances here rather than operating on `*os.File`s directly.
</blockquote></details>

This is to write to the artifacts file that is being written to in query_comparison_util - so we have the entire trace of tables, queries etc that are being used for our test  - we are writing to to this file directly in other parts of the code so I think we should keep it consistent.


<!-- Sent from Reviewable.io -->

rachitgsrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jlinder, @michae2, @renatolabs, and @rytaft)

___
*[`pkg/internal/workloadreplay/workloadreplay.go` line 120 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWa4flAc94nxpHKR4A:-N_JsyB82dzECweWGtui:b-7tg4cd) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/internal/workloadreplay/workloadreplay.go#L120)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Might be worth formatting these queries to make them more palatable and easier to adapt when we have to.
</blockquote></details>

acked


<!-- Sent from Reviewable.io -->

renatolabs reviewed (approved):
As discussed, now that `log.Fatal` was removed, I think we can merge as-is and make any further improvements as follow-up PRs.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-:-N_KHGcF7FsJRVC3zCuh:b-avv4vg)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @jlinder, @michae2, and @rytaft)



<!-- Sent from Reviewable.io -->

rachitgsrivastava reviewed (commented):
@renatolabs @rytaft  I have tested the changes - we can merge the pr now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-:-N_LuFL4951DN5w7KTIO:b-1mqe5u)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @jlinder, @michae2, @renatolabs, and @rytaft)

___
*[`pkg/cmd/roachtest/tests/admission_control_elastic_backup.go` line 101 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/95390#-NZWVJkX3-7m_1fQjYnJ:-N_LuAYi8L3Mvi_M4rvM:b-lsfasp) ([raw file](https://github.com/cockroachdb/cockroach/blob/209179769665a650b41c7fba57548b402d6b0287/pkg/cmd/roachtest/tests/admission_control_elastic_backup.go#L101)):*
<details><summary><i>Previously, rachitgsrivastava wrote…</i></summary><blockquote>

sorry what variable are you talking about here?
</blockquote></details>

This has been changed too.


<!-- Sent from Reviewable.io -->

rachitgsrivastava commented:
bors r+

rachitgsrivastava commented:
bors r+

renatolabs commented on .github/CODEOWNERS:
```diff
@@ -477,6 +477,7 @@
 /pkg/internal/rsg/           @cockroachdb/sql-queries
 /pkg/internal/sqlsmith/      @cockroachdb/sql-queries
 /pkg/internal/team/          @cockroachdb/test-eng
+/pkg/internal/workloadreplay/          @cockroachdb/test-eng-noreview
```
I think we want `@cockroachdb/test-eng` not `@cockroachdb/test-eng-noreview` here, no?
